### PR TITLE
Make CUDAWrappers::MatrixFree compile with 64-bit indices

### PR DESCRIPTION
**Pull request by masterleinad**
We just have to swap a few `unsigned int` and `types::global_dof_index`.

### CODE REVIEW

File: include/deal.II/matrix_free/cuda_matrix_free.h [hunk 1]
```diff
@@ -259,14 +259,18 @@ namespace CUDAWrappers
      */
     std::vector<Number *> JxW;
 
-    // Constraints
-    unsigned int *              constrained_dofs;
+    // Pointer to the constrained degrees of freedom.
+    types::global_dof_index *constrained_dofs;
+
+    // Mask deciding where constraints a set on a given cell.
```
Comment (davydden) [inline include/deal.II/matrix_free/cuda_matrix_free.h]: are set?

Comment (masterleinad) [reply include/deal.II/matrix_free/cuda_matrix_free.h]: Yes, of course. Fixed.

**Comment by masterleinad** [comment]
@davydden I had to fix a few more places for the tests to pass.

**Comment by masterleinad** [comment]
> Looks good. Have you run all the tests? Do you know if it's really feasible to run a problem with more than 4 billions unknowns on one card?

All the tests run successfully. Probably, it's not feasible to run such large problems on one card and I didn't check that this really works.
This PR only makes sure that we can compile with `DEAL_II_WITH_64BIT_INDICES=ON` rather than restricting `CUDA` support to `DEAL_II_WITH_64BIT_INDICES=OFF`.

**Comment by Rombur** [comment]
OK let's merge this. The tester does not check CUDA code anyway.

File: include/deal.II/matrix_free/cuda_hanging_nodes_internal.h [hunk 2]
```diff
@@ -282,7 +283,7 @@ namespace CUDAWrappers
 
       std::vector<types::global_dof_index> neighbor_dofs(dofs_per_face);
 
-      std::vector<types::global_dof_index> lex_face_mapping =
+      const auto lex_face_mapping =
         FETools::lexicographic_to_hierarchic_numbering<dim - 1>(
```
Comment (bangerth) [inline include/deal.II/matrix_free/cuda_hanging_nodes_internal.h]: The change is not wrong, but out of curiosity: what was the reason to change this? Isn't the type always `std::vector<unsigned int>`? (I.e., it was clearly wrong before.)

Comment (masterleinad) [reply include/deal.II/matrix_free/cuda_hanging_nodes_internal.h]: Yes, it is. At that point, I was just happy to let the compiler choose the right type and it is shorter.
Call it laziness. I am not sure if we have a coding standard in this regard.

Comment (bangerth) [reply include/deal.II/matrix_free/cuda_hanging_nodes_internal.h]: No. I typically err on the side of being explicit unless the type is obvious (e.g., for cells). But I'm happy to leave as is :-)

**Pull request merged**